docs: add november contributions post (#37)
e18e/e18e: docs/blog/november-contributions-showcase.md

Often, this has been achieved by using a micro-utility dependency or by repeating a very similar function to everyone else who needed this logic.

This is a really good example of something which is too big to be copied everywhere, but too small to justify its own package.

So what do we do in these cases? We group the common case transforms and use _that_ as a library.

That is exactly what Fabio has done with [kasi](https://github.com/fabiospampinato/kasi)!

kasi is a great collection of functions for transforming between various cases (pascal, camel, snake, etc). Instead of pulling in many micro-packages, or repeating code, we can now use this.

Great work as always, Fabio! :pray:

### unicode-segmenter

As part of the bluesky contributions mentioned later in this post, some of the community noticed a large chunk of the JS bundle was taken up by [graphemer](https://github.com/flmnt/graphemer).

This library is used to split or count the number of grapheme clusters in a string (i.e. the visual symbols we would perceive as individual characters). It does a good job at this but comes in at a whopping 812KB!

> Fortunately, [@hyeseong.kim](https://bsky.app/profile/hyeseong.kim) came along and built [unicode-segmenter](https://github.com/cometkim/unicode-segmenter) :tada: This library achieves the same functionality at a fraction of the size (250KB).

Reviewer: How do you get this number (250KB) ? The same grapheme functionality is only takes 17313 bytes even without minification in the latest version (v0.11.2)

Reviewer: You can check it from [bundlejs](https://bundlejs.com/?q=unicode-segmenter%2Fgrapheme&treeshake=%5B*%5D) (![](https://deno.bundlejs.com/badge?q=unicode-segmenter/grapheme&treeshake=[*]))
More detail stats is on README: https://github.com/cometkim/unicode-segmenter?tab=readme-ov-file#js-bundle-stats

Author: interesting, both [pkg-size](https://pkg-size.dev/unicode-segmenter) and [npm](https://www.npmjs.com/package/unicode-segmenter) say 250KB

maybe because I'm looking at overall package size rather than just the grapheme entry point?

Reviewer: Oh ok that means NPM package size, not bundle size? That makes sense, but 250KB seems to large still, let me check a quick

Reviewer: The package is bloated mostly by commonjs entries and its sourcemap :sweat_smile: 

It is not the actual bundle size anyway, the exact size of the graphemer or Intl alternative is much smaller due to the tree-shaking.

<details>

<summary> npm pack </summary>

```
npm notice
npm notice 📦  unicode-segmenter@0.11.2
npm notice Tarball Contents
npm notice 2.4kB _emoji_data.cjs
npm notice 2.1kB _emoji_data.cjs.map
npm notice 618B _emoji_data.d.ts
npm notice 1.7kB _emoji_data.js
npm notice 11.7kB _general_data.cjs
npm notice 11.5kB _general_data.cjs.map
npm notice 711B _general_data.d.ts
npm notice 11.1kB _general_data.js
npm notice 13.0kB _grapheme_data.cjs
npm notice 15.0kB _grapheme_data.cjs.map
npm notice 3.0kB _grapheme_data.d.ts
npm notice 14.0kB _grapheme_data.js
npm notice 1.3kB _incb_data.cjs
npm notice 838B _incb_data.cjs.map
npm notice 311B _incb_data.d.ts
npm notice 589B _incb_data.js
npm notice 1.1kB LICENSE
npm notice 13.6kB README.md
npm notice 2.0kB core.cjs
npm notice 3.9kB core.cjs.map
npm notice 3.7kB core.d.ts
npm notice 2.8kB core.js
npm notice 1.5kB emoji.cjs
npm notice 1.3kB emoji.cjs.map
npm notice 647B emoji.d.ts
npm notice 937B emoji.js
npm notice 1.6kB general.cjs
npm notice 1.5kB general.cjs.map
npm notice 666B general.d.ts
npm notice 979B general.js
npm notice 5.9kB grapheme.cjs
npm notice 13.3kB grapheme.cjs.map
npm notice 2.2kB grapheme.d.ts
npm notice 9.6kB grapheme.js
npm notice 1.2kB index.cjs
npm notice 339B index.cjs.map
npm notice 117B index.d.ts
npm notice 131B index.js
npm notice 3.4kB intl-adapter.cjs
npm notice 4.5kB intl-adapter.cjs.map
npm notice 2.1kB intl-adapter.d.ts
npm notice 3.1kB intl-adapter.js
npm notice 402B intl-polyfill.cjs
npm notice 707B intl-polyfill.cjs.map
npm notice 11B intl-polyfill.d.ts
npm notice 342B intl-polyfill.js
npm notice 1.1kB licenses/unicode-segmentation_MIT.txt
npm notice 4.3kB package.json
npm notice 1.7kB src/_emoji_data.js
npm notice 11.1kB src/_general_data.js
npm notice 14.0kB src/_grapheme_data.js
npm notice 589B src/_incb_data.js
npm notice 2.8kB src/core.js
npm notice 937B src/emoji.js
npm notice 979B src/general.js
npm notice 9.6kB src/grapheme.js
npm notice 131B src/index.js
npm notice 3.1kB src/intl-adapter.js
npm notice 342B src/intl-polyfill.js
npm notice 3.0kB src/utils.js
npm notice 2.4kB utils.cjs
npm notice 4.3kB utils.cjs.map
npm notice 2.1kB utils.d.ts
npm notice 3.0kB utils.js
npm notice Tarball Details
npm notice name: unicode-segmenter
npm notice version: 0.11.2
npm notice filename: unicode-segmenter-0.11.2.tgz
npm notice package size: 58.3 kB
npm notice unpacked size: 239.0 kB
npm notice shasum: f15eca105936dd7f48075b9b7c0d915066d58f6a
npm notice integrity: sha512-ssdWyVdO9kfmt[...]R4Kdidgdw1ThA==
npm notice total files: 64
npm notice
unicode-segmenter-0.11.2.tgz
```

</summary>

Reviewer: ESM entries seems to be duplicated. It uses the source as is without any building, so there is no reason to include `*.js` once more in the root. Good catch

Author: in the blog posts so far, we've been using _install size_, so we should stick with that here too

it sounds like you will be able to make it smaller though. it is still pretty small compared to grapheme anyway 

I might still mention bundle size just to be clear though

Reviewer: If I build CommonJS entries in a more compact format, they don't even need a source map as they are fully readable. Thus, the total package size can be reduced to 135KB !

![image](https://github.com/user-attachments/assets/10f94464-6a02-4bee-9b4e-68f84691cdd1)

Reviewer: Of course it will be even much better if it uses `require(ESM)` :joy:

Reviewer: https://github.com/cometkim/unicode-segmenter/releases/tag/unicode-segmenter%400.11.3

Really great work here. This helped us reduce bundle sizes in a whole bunch of projects.

### sonda

As part of the cleanup effort in the e18e space, it is a common need that we want to analyze the bundle of a project. Usually, we do this to find duplicate dependencies, oversized dependencies and so on.

In the wild, there are not many tools to do this analysis and do it accurately. This has left us with often manually inspecting bundles and the process taking a lot longer than it should.

All good, though, [Filip](https://bsky.app/profile/filipsobol.bsky.social) has come to the rescue with his new tool: [sonda](https://github.com/filipsobol/sonda/tree/main/packages/sonda)!

> Sonda is a bundle analyzer which works with all popular bundlers by leaning on sourcemaps to determine the bundle contents. This works great and has already helped us with many investigations into footprint reduction of some projects :pray:

Reviewer 2: > This [...] already helped us with many investigations into footprint reduction of some projects

Maybe I'm working against myself here, but is it true? :sweat_smile:

Author: i've been using it on vite's bundle recently and a few others to catch duplicates etc

I've seen a few others using it too, but its difficult to say for sure what the actual usage is

maybe I should reword that it is _starting_ to help us?

Reviewer 2: Nah, I just wanted to make sure that this is not made up just to make it sound more exciting :wink:

> Sonda is a bundle analyzer which works with all popular bundlers by leaning on sourcemaps to determine the bundle contents. This works great and has already helped us with many investigations into footprint reduction of some projects :pray:

Reviewer 2: ```suggestion
Sonda is a bundle analyzer that works with all popular bundlers and accurately determines the bundle contents by leaning on sourcemaps. This works great and has already helped us with many investigations into footprint reduction of some projects :pray:
```

## Other news

### deoptexplorer

More of a "you should know about this" than news - [deoptexplorer](https://github.com/microsoft/deoptexplorer-vscode) is an extremely useful VSCode plugin to help you find deoptimisations in your code.

In the community, we have started using it heavily to find performance bottlenecks in many different popular tools.

Big thanks to microsoft for building this! Check it out and maybe you'll find some performance improvements :rocket:

### `require(esm)` unflagged in Node 22.x

In Node 22.x, the ability to `require(...)` ES modules has been unflagged and enabled by default!

This is huge news and should open many doors for us being able to migrate libraries to ESM-only. We should also be able to gradually start removing dual-package publishing from various popular packages and go all in on ESM.

Read more about this in [Node's blog post](https://nodejs.org/en/blog/release/v22.12.0#requireesm-is-now-enabled-by-default).

Excellent work by [@joyeecheung.bsky.social](https://bsky.app/profile/joyeecheung.bsky.social) as always :pray:

### vite 6.0

TODO

ref: https://github.com/vitejs/vite/blob/main/packages/vite/CHANGELOG.md#performance

> 

Reviewer 3: there's also the switch to tinyglobby and chokidar 4, which is up above in the breaking changes section, and which reduce bundle size

In the community, we have started using it heavily to find performance bottlenecks in many different popular tools.

Big thanks to microsoft for building this! Check it out and maybe you'll find some performance improvements :rocket:

### `require(esm)` unflagged in Node 22.x

In Node 22.x, the ability to `require(...)` ES modules has been unflagged and enabled by default!

This is huge news and should open many doors for us being able to migrate libraries to ESM-only. We should also be able to gradually start removing dual-package publishing from various popular packages and go all in on ESM.

Read more about this in [Node's blog post](https://nodejs.org/en/blog/release/v22.12.0#requireesm-is-now-enabled-by-default).

Excellent work by [@joyeecheung](https://bsky.app/profile/joyeecheung.bsky.social) as always :pray:

### vite 6.0

A big milestone this month - vite 6.0 was released!

This new release has a huge number of changes and, in the context of e18e, has a great list of performance improvements many of our community worked on.

> Two big changes are the upgrade of [chokidar](https://github.com/paulmillr/chokidar) to 4.x, and the migration to [tinyglobby](https://github.com/SuperchupuDev/tinyglobby). These two alone reduced the bundle size greatly :broom:

Reviewer 4: Sorry for the late review. I just noticed this part but Vite 6 had reverted to using chokidar v3 still.
